SparseGrids/gridtestExternalTests.cpp

```diff
@@ -241,25 +241,40 @@ TestResults ExternalTester::getError(const BaseFunction *f, TasGrid::TasmanianSp
     int num_dimensions = f->getNumInputs();
     int num_outputs = f->getNumOutputs();
     int num_points = grid.getNumPoints();
-    if (type == type_integration or type == type_nodal_interpolation){
+    if (type == type_integration or type == type_nodal_interpolation or type == type_nodal_differentiation){
+        int num_entries = (type == type_nodal_differentiation) ? num_dimensions * num_outputs : num_outputs;
+        std::vector<double> y(num_entries);
+        std::vector<double> r(num_entries, 0.0);
         auto points = grid.getPoints();
-        auto weights = (type == type_integration) ? grid.getQuadratureWeights() : grid.getInterpolationWeights(x);
+        std::vector<double> weights;
+        if (type == type_integration)
+            weights = grid.getQuadratureWeights();
+        else if (type == type_nodal_interpolation)
+            weights = grid.getInterpolationWeights(x);
+        else
+            weights = grid.getDifferentiationWeights(x);
 
-        std::vector<double> y(num_outputs);
-        std::vector<double> r(num_outputs, 0.0);
-//      Sequential version: integration
+        // Sequential version
         for(int i=0; i<num_points; i++){
             f->eval(&(points[i*num_dimensions]), y.data());
-            for(int k=0; k<num_outputs; k++) r[k] += weights[i] * y[k];
+            for(int k=0; k<num_entries; k++) {
+                int widx = (type == type_nodal_differentiation) * (i * num_dimensions + k % num_dimensions) +
```

Comment:
The body of the for-loop is absolutely DISGUSTING!

Before you fix it, you have to tell me at least two reasons why this should never be written.

Reply:
Changed (according to our in-person discussion).